text/0000-pub-sub.md

```diff
@@ -0,0 +1,171 @@
+
+- By configuring specific event subscriptions for each tenant it is possible to implement specific cases of inter-app coordination.
+
+- The `mod-pubsub` module is implemented using a message queue (Kafka) which will provide the necessary persistence.
+
+#### High-level solution structure
+There is a global FOLIO interface `pubsub` defined in the scope of this solution which provides APIs to ensure the functioning of the subsystem:
+- API for event type registration by publishers and subscribers. This API must be called by back-end modules.
+  - `POST /pubsub/eventtypes` - to register an event type 
+  - `DELETE /pubsub/eventtypes` - to remove event type registration
+  - `POST /pubsub/eventtypes/declare/publisher` - to declare a publisher for a set of event types
+  - `DELETE /pubsub/eventtypes/declare/publisher` - to remove publisher declaration
+  - `POST /pubsub/eventtypes/declare/subscriber` - to declare a subscriber for a set of event types
+  - `DELETE /pubsub/eventtypes/declare/subscriber` - to remove subscriber declaration
+- API to retrieve existing registered event types.
+  - `GET /pubsub/eventtypes` - returns all registered event types
+  - `GET /pubsub/eventtypes/{EVENT_TYPE}` - returns an event descriptor for a particular event type
+- API to retrieve the registered subscriptions for a specified tenant and event type
+  - `GET /pubsub/eventtypes/{EVENT_TYPE}/subscribers` - returns all subscribers of the event type provided in the path variable
+  - `GET /pubsub/eventtypes/{EVENT_TYPE}/publishers` - returns all publishers of the event type provided in the path variable
+- API used by publishers to send events. This API must be called by back-end modules which act as publishers
+  - `POST /pubsub/publish`
+- API to retrieve the activity history performed by the event manager. A correlationID can optionally be provided to limit context.
+  - `GET /pubsub/history`
+  - `GET /pubsub/history?corellationId={corellationId}`
+
+It is allowed for publishers to provide callback URLs to receive feedback from `mod-pubsub` module in cases when:
+- Subscriber did not receive an event within the specified period of time.
+- All subscribers deleted their subscriptions on a particular event type registered by this publisher.
+ 
+
+There is a set of requirements that must be followed by a module to provide “Subscriber endpoints” to consume events. The basic set of requirements:
+- HTTP Method - POST
+- A set of required HTTP header. [_To be defined further if it is needed._]
+- A payload structure - JSON which represents an event. See [Event data structure](#Event data structure) 
+
+Reference implementations should be provided that allow a back-end module to become a publisher or a subscriber respectively to simplify development.
+  
+A dedicated module, `mod-pubsub`, must implement and encapsulate all logic related to event sending and delivery. All interactions between this module and back-end business modules will be done via HTTP.
+
```

Comment:
How is high availability addressed in in `mod-pubsub`? Do we run multiple modules, do they need to be clustered, etc.?

How is high availability addressed in the subscribers? If we have 10 `mod-circulation` modules running and they all subscribe to the same set of events, will that be a problem?

Will a pubsub mock framework be available for use by publishing/subscribing modules for unit tests?

Is it a problem if modules don't deregister themselves as publishers/subscribers as in the case of a module crash? Below it says that modules must deregister on tenant disablement, what if that fails?

Reply:
High availability consideration are certainly something to take into account as part of the implementation.  This would be reflected in how the underlying Kafka/Zookeeper are configured, but also some operational choices relating to how multiple module instances (horizontal scaling) are configured in the system. 

The test harness is not specifically detailed in this proposal, but would come up as part of implementation. I think it would be reasonable to provide mock responses for publisher and subscriber registration calls, as well as mock responses for publishing calls, for the purposes of unit tests. The subscriber handler endpoints are just normal endpoints and can provide normal unit tests. Additionally, integration tests will also be needed.

As to the deregistration question... In the case of the publishing module, I don't see an issue. If the module is removed anyways or crashed it just won't publish. It can be cleaned up again later. For the subscriber, there would be a need for the pubsub module to be resilient in its attempt to push a message. If the subscriber's endpoint is incorrect or unavailable it should be able to raise an error and continue.  For tenant disablement, it raises a question about how to proceed, but ultimately the problem sits outside of the pubsub module itself. It would be another process that takes care of  disabling tenant. The pubsub module will do its part when called upon and respond with an error if it can't. How that is dealt with is outside the pubsub module.